Memory snapshot harness lines

xbauch opened the pull request:
Before the user had to run the CBMC first to find out the location where they wanted to start their snaphot. This PR adds to option to specify the code line.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -69,6 +76,66 @@ void memory_snapshot_harness_generatort::handle_option(
+      const auto &goto_function = entry.second;
+
+      if(!goto_function.body_available())
+        continue;
```
That's actually unnecessary, the `std::find_if` will do the right thing either way.

xbauch replied:
I don't understand. I call `find_if` on the instructions inside `goto_function.body`. If the body is not `body_available` is it still safe to do?

tautschnig replied:
It is safe (`body_available()` just tests emptiness), but arguably that could change at some point in the future.

xbauch replied:
Ok, I removed it but left a comment.

tautschnig commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -69,6 +76,66 @@ void memory_snapshot_harness_generatort::handle_option(
+
+      if(it != goto_program.instructions.end())
+      {
+        entry_function_name = entry.first; //TODO - move elsewhere
```
I don't think I understand this `TODO`.

xbauch replied:
Sorry, should have been removed before.

tautschnig commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -69,6 +76,66 @@ void memory_snapshot_harness_generatort::handle_option(
+        goto_program.instructions.end(),
+        [this](const goto_programt::instructiont &instruction) {
+          return safe_string2unsigned(id2string(
+                   instruction.source_location.get_line())) >= *line_number;
```
But then the same line number can occur in multiple files?

xbauch replied:
Well `find_if` should return the first location with "high enough" line number. Is this still relevant if we insist on having the user specify the file as well?

tautschnig replied:
If the user can/has to specify the file then that's sorted, thanks!

tautschnig commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -69,6 +76,66 @@ void memory_snapshot_harness_generatort::handle_option(
+      const auto &goto_program = goto_function->second.body;
+      if(!location_number.has_value())
+        return goto_program.instructions.begin();
+      if(goto_program.instructions.size() > *location_number)
```
Why is this condition necessary?

xbauch replied:
Because I'm advancing the iterator by `*location_number` position and I don't want to get past the `end()`.

tautschnig replied:
Ah, sorry, I missed that. But also this isn't how the `location_number` member of instructions works: these location numbers are guaranteed to be unique across _all_ goto functions, and monotonically increasing in a single function. Therefore a location number could well be larger than the size of a single function. At the same time, there could be gaps, so locations in a function could legitimately be numbered "7, 10, 15" instead of "1, 2, 3."

xbauch replied:
I give up. What about `find_if` to get the right instruction? That should be safe right?

tautschnig replied:
Yes, I'm afraid you will have to do some form of search (technically binary search would be faster, but quite likely no one will notice the cost of linear search here).

xbauch replied:
But would binary search on a `std::list` really be faster? I thought that random access there is linear in the `distance` of the jump.

tautschnig replied:
That's a question of how expensive the comparison operation is. Most likely you are right in that binary search would not add any value (or may indeed be more expensive).

tautschnig commented on src/goto-harness/memory_snapshot_harness_generator.h:
```diff
@@ -34,6 +35,7 @@ Author: Daniel Poetzl
   "--initial-location <func[:<n>]>\n"                                          \
   "                              use given function and location number as "   \
   "entry\n                              point\n"                               \
+  "--initial-line <n>            use given line as entry point\n"              \
```
Can we actually use `file:line`? I think the line number alone is too ambiguous. You might actually end up picking an instruction from some system header file!

xbauch replied:
That's a very good idea. Added in the last commit. Which will also be squashed.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 1f0d7e9c2e3c8a1333545d729edb0626ebce28d8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/105415411
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: e2598d2d242dcba66c2226226cc787c82e232c4d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/105422508
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 948b5752df40dcd8bca1af57a740dc503689faf0).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/105424064
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

martin-cs commented on src/goto-programs/goto_program.h:
```diff
@@ -588,28 +588,6 @@ class goto_programt
     return t;
   }
 
-  optionalt<const_targett> get_target(const unsigned location_number) const
```
I take it this was only used by goto-harness.

xbauch replied:
It was used by the option `--initial-location` to find the instruction with this `location_number`. Now that we have two ways for the user to specify the entry point I generalised the search for the right instruction and thus `get_target` was no longer needed.

martin-cs replied:
OK; sounds good.

martin-cs reviewed (commented):
Seems like a reasonable feature.  Will approve once @tautschnig 's comments have been resolved.

hannes-steffenhagen-diffblue commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -61,6 +63,24 @@ void memory_snapshot_harness_generatort::handle_option(
   {
     variables_to_havoc.insert(values.begin(), values.end());
   }
+  else if(option == "initial-file-line")
```
:bike:  :derelict_house:  maybe `--initial-source-location`?

xbauch replied:
Done.

hannes-steffenhagen-diffblue commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -61,6 +63,24 @@ void memory_snapshot_harness_generatort::handle_option(
+      require_exactly_one_value(option, values);
+
+    std::vector<std::string> start;
+    split_string(initial_file_line, ':', start, true);
```
`split_string` has a special cased version for working with 2 values, maybe worth considering?

xbauch replied:
Done.

hannes-steffenhagen-diffblue commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -61,6 +63,24 @@ void memory_snapshot_harness_generatort::handle_option(
+        "invalid initial location specification", "--initial-file-line");
+    }
+
+    initial_file_name = start.front();
```
:bulb:  might be worth wrapping these two in a struct to keep (and make obvious) the invariant that either both or neither of these are set

xbauch replied:
@hannes-steffenhagen-diffblue This required a little bigger change: the last 7 commits. Please have a look if its still acceptable.

hannes-steffenhagen-diffblue reviewed (approved):
Looks good to me overall, not even really nitpicks just suggestion/questions

hannes-steffenhagen-diffblue commented on src/goto-harness/memory_snapshot_harness_generator.h:
```diff
@@ -69,6 +69,46 @@ class memory_snapshot_harness_generatort : public goto_harness_generatort
+    irep_idt function_name;
+    optionalt<goto_programt::const_targett> start_instruction;
+
+    void initialize_via_goto(
```
I'd be more comfortable if these were builder/constructor functions, i.e. `entry_locationt make_from_goto_location(...);`, `entry_locationt make_from_source_location`

xbauch replied:
Good idea: done.

hannes-steffenhagen-diffblue commented on src/goto-harness/memory_snapshot_harness_generator.h:
```diff
@@ -69,6 +69,46 @@ class memory_snapshot_harness_generatort : public goto_harness_generatort
+    irep_idt function_name;
+    optionalt<unsigned> location_number;
+
+    entry_goto_locationt(const std::string &cmdl_option);
```
`explicit` - also, I'm personally not a huge fan of "do work" constructors, i.e. this one parses from string, I'd prefer if it was a function that said it did that in its name

xbauch replied:
Done + constructors are now trivial.

hannes-steffenhagen-diffblue commented on src/goto-harness/memory_snapshot_harness_generator.h:
```diff
@@ -69,6 +69,46 @@ class memory_snapshot_harness_generatort : public goto_harness_generatort
+    irep_idt file_name;
+    unsigned line_number;
+
+    entry_source_locationt(const std::string &cmdl_option);
```
also `explicit` and the above comment

xbauch replied:
Done.

hannes-steffenhagen-diffblue commented on src/goto-harness/memory_snapshot_harness_generator.h:
```diff
@@ -69,6 +69,46 @@ class memory_snapshot_harness_generatort : public goto_harness_generatort
+      const entry_source_locationt &entry_source_location,
+      const goto_functionst &goto_functions);
+
+    bool valid() const
```
Ideally it shouldn't be possible to construct this in an invalid state, if it can be avoided

xbauch replied:
Done.

hannes-steffenhagen-diffblue commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -378,3 +378,113 @@ void memory_snapshot_harness_generatort::generate(
+
+  if(start.size() == 2)
+  {
+    location_number = optionalt<unsigned>(safe_string2unsigned(start.back()));
```
this should be `string2optional_unsigned` instead, and you should check the result... if this function returns `nullopt` it means its argument wasn't a valid integer, which is something you don't want to silently ignore

(with `safe_string2unsigned` you're basically asserting that this conversion will always succeed which is not correct for user supplied strings)

xbauch replied:
Done.

hannes-steffenhagen-diffblue commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -378,3 +378,113 @@ void memory_snapshot_harness_generatort::generate(
+  }
+
+  file_name = initial_file_string;
+  line_number = safe_string2unsigned(initial_line_string);
```
same as above

xbauch replied:
Done.

hannes-steffenhagen-diffblue commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -66,21 +66,43 @@ void memory_snapshot_harness_generatort::validate_options(
+      "choose either source or goto location to specify the entry point",
+      "--initial-source/goto-location");
+  }
+
```
I think this needs a check that only one of goto location and source location is set?

xbauch replied:
It's there somewhere (line 69).

Commit:
Add and modify regression tests

To test the new entry point selection and reflect the renaming of the option.

Commit:
Add source location option

And unify the option names.

Commit:
Introduce wrappers for entry location

Two kinds of user specification: either via goto location or via source line. We
wrap these in entry_goto/source_locationt and build the production of the final
entry_locationt.

Commit:
Simply handling the options

We only store the strings here.

Commit:
Reflect the existence of wrappers

Most of the sanity checks are in the wrappers and the `get_target` function is
now redundant.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 0249ba059cebb09eba226f99f444c04de3779505).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106287968
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.
